RFC-0001-Agile Coretime.md

```diff
@@ -0,0 +1,291 @@
+
+The Broker-chain aims to sell some `BULK_TARGET` of Cores, up to some `BULK_LIMIT`. It makes this sale in a single batch `LEADIN_PERIOD` prior to the beginning of the period being sold. The Broker chain publishes a price for this batch of sales for the `BULK_TARGET` period prior to the sale execution.
+
+Accounts may call a `purchase` function with the appropriate funds in place to have their funds reserved and signal an intention to purchase Bulk Coretime for the forthcoming period. One account may have only one pending purchase. Any number of accounts may attempt to `purchase` Bulk Coretime for the forthcoming period, but the order is recorded.
+
+If there are more purchase requests than available cores for purchase in this period, then requests are processed in incoming order. Any additional purchase orders are carried over but marked as such. A purchase is only cancellable if it was carried over.
+
+When a region of Bulk Coretime is initially issued through this purchase, the price it was purchased for is recorded, in addition to the beginning and end Relay-chain block numbers to which it corresponds.
+
+The Broker-chain SHALL record this information in a storage double-map called Regions, keyed first by the current bulk `SaleIndex` (a `u16` starting at zero and incrementing with each sale), then secondarily by a `RegionId`. It shall map into a value of `RegionRecord`:
+
+```rust
+type SaleIndex = u16;
+struct RegionId {
+    core: CoreIndex,  // A `u16`.
+    begin: Timeslice, // A `u16`.
+}
+struct RegionRecord {
+    owner: AccountId,
+    end: Timeslice,
+    price: Option<Balance>,
+    allocation: Option<Vec<ParaId>>, // begins set to `None`
+}
+```
+
+This map functions essentially as a linked list. With one region's `end` field functioning as the next region's key's `begin` field. It is keyyed by the sale index in order to allow the following sale period's Coretime to be manipulated during the `LEADIN_PERIOD` prior to it becoming allocatable.
+
+An additional storage map is maintained to keep the "heads" of this linked list. It is called `NextRegion` and it maps `CoreIndex` to `Timeslice`, to indicate the earliest stored region of the given core.
+
+Notably, if a region is split or transferred, then the `price` is reset to `None`, which has the effect of disallowing a price-increase-capped renewal.
+
+The Broker-chain provides feedback to the Relay-chain on which `ParaId` sets should be serviced on which cores, and does so as they change. Each block, the Broker-chain checks if the period of a `Timeslice` has elapsed. If so, it checks to see if any cores have a newly active `RegionRecord` value in the `Regions` map. If they do, it MUST notify the Relay-chain of the new responsibilities of the relevant `core`. In this case, it MUST remove the item from the `Regions` map and update the `NextRegion` map so it maps the relevant core to the value of removed record's `end` field.
```

Comment:
Given that parachains have bounded execution, it may be necessary to split this over a few blocks when dealing with large numbers of cores (e.g. 500-1000)

Reply:
Indeed. I don't see any great problem with having multiple `reallocate_core_soon` messages which can be trivially split over multiple blocks if needed. Most of the time I'd expect the changes not be aligned with each other (i.e. only one or two cores changing at any given Timeslice boundary) and therefore the typical case would be pretty trivial.